### DISCUSSION TITLE
Using global variables in frontmatter?

### DISCUSSION THREAD

**Discussion by RosemaryOrchard** [ttscoff/bunch-docs#243] [Q&A]
I have a number of global variables - things like the UUID of various Macs, etc. I've been trying to use these in the frontmatter with `ignore` and `ignore if`, and while the bunches aren't showing up, they're not showing up on any machines! I've tried:

```yaml
ignore: workMac
```

And

```yaml
ignore if: ${workMac}
```

(Both ways with and without the if in the ignore if.)

Unfortunately, I've hit the end of the obvious things to try that I can think of. Something that may be important: I'm using these in folder and tag frontmatter, but I seem to have the same problem directly in Bunches. 

Might it be because the global variables are set from a bunch in a folder? That's the only thing I've got left to try! (And shall do so shortly, asking questions is always a useful troubleshooting exercise!)

Edit: sadly moving my bunch that sets variables to the main folder, and then running it manually, did not magically solve this automation conundrum. Darn

**Comment by ttscoff** [comment]
Just to be certain, if you run `(log "${workMac}")` does the UUID show 
up without anything extraneous?

I'm going to have to refresh myself on the conditional logic in 
frontmatter. I know `if uuid is ${workMac}` should work in a logic 
statement within a Bunch, and it _should_ work the same in front matter, 
but I'll have to dig and see if there's some reason it wouldn't.

**Comment by ttscoff** [comment]
Turns out I wasn't processing variables at all before frontmatter handling, as usually variables wouldn't be assigned yet (I neglected to consider global variables). Beta 161 is out now and corrects this issue, handling known global variables before processing any conditions in frontmatter. Note that the global variable has to be set _before_ the Bunch in question loads, it won't register a value within the Bunch prior to parsing the frontmatter.

Hope that solves it, but let me know!

**Comment by ttscoff** [comment]
Oh, btw, tested with `ignore if: uuid is ${workMac}` and `ignore if: ${workMac}`. The brackets are generally a good idea for clarity, both in parsing and in general readability.

**Comment by RosemaryOrchard** [comment]
> Just to be certain, if you run `(log "${workMac}")` does the UUID show up without anything extraneous?

Yup! Definitely working. The variables are set regularly through a scheduled bunch.

> Turns out I wasn't processing variables at all before frontmatter handling, as usually variables wouldn't be assigned yet (I neglected to consider global variables)

I did wonder if it were something like that, but it would make sense that variables wouldn’t be considered. Thank you for adding it!

> Oh, btw, tested with `ignore if: uuid is ${workMac}` and `ignore if: ${workMac}`. The brackets are generally a good idea for clarity, both in parsing and in general readability.

Agreed, I tried without as a test—it makes sense to use them.

I appreciate the new feature! 😁

**Comment by RosemaryOrchard** [comment]
I just wanted to follow up on this, and say that recently I have noticed that these do not work in folder/tag front matter (anymore, I am sure they did).

I now also have the added wrinkle of using a personal machine as a work machine (at times). So here's the big picture:

I have a `Variables.bunch` file which runs periodically, and on startup. This contains:
```
global isWorkMac = $ if [ $(system_profiler SPHardwareDataType | awk '/UUID/ { print $3; }') = '${workMac}' ] || [ $(whoami) = '${workUser}' ] ; then echo true; else echo false; fi
```

Then in `Podcasting/folder.frontmatter` (and some others), as well as `@podcast.frontmatter` (in the root of my Bunch folder) the front matter contains:
```
ignore: `if ${isWorkMac} is true`
```

The following `test.bunch` however doesn't show up, and if I comment out the `ignore` in the front matter, then it shows up and logs exactly what I expect (and the notify outputs too):
```
---
title: A Test
ignore: `if ${isWorkMac} is true`
toggles: false
---
(notify Is Work Mac: ${isWorkMac})
(log "workMac: ${workMac}")
(log "workUser: ${workUser}")
(log "isWorkMac: ${isWorkMac}")
```

I'm not sure what to try next, but figured it was about time to ask for help! 😁

**Comment by RosemaryOrchard** [comment]
Aha, I tracked it down!

Because I had `ignore if: ${isWorkMac}` (I tried that as well as the ignore mentioned above) in my `@podcast.frontmatter` file, _and_ in my `Podcasting/folder.frontmatter` (where some but not _all_ bunches are tagged with `podcast`, the folder front matter was being ignored.

If I use
```
ignore: `if ${isWorkMac} is true`
```
and the `ignore if: ${isWorkMac}` in another, then there are no problems and it all works as expected.

The good news is I can fix the problem for myself by remembering that all the bunches tagged podcast are already hidden by their folder.

**Comment by ttscoff** [comment]
Well, this got complicated. I'm not immediately wrapping my head around 
why this isn't working or why your solution does. I'll have to dig in, 
test, and figure out the pattern to understand.

**Reply by RosemaryOrchard** [reply]
I suspect it's to do with the way the front matter is being combined along with a dash of "what is truthy/falsey" (though I'm sure you have concluded that as well). 

As I mentioned, I've solved the issue for myself so it's not urgent, but as I found the reason why it seemed worth telling you!
